atc/db/build_in_memory_check.go

```diff
@@ -0,0 +1,445 @@
+	pipeline, found, err := b.Pipeline()
+	if err != nil {
+		return nil, fmt.Errorf("failed to find pipeline: %w", err)
+	}
+	if !found {
+		return nil, errors.New("pipeline not found")
+	}
+
+	return pipeline.Variables(logger, secrets, varSourcePool)
+}
+
+func (b *inMemoryCheckBuild) SaveEvent(ev atc.Event) error {
+	if !b.running {
+		panic("not a running in-memory-check-build")
+	}
+
+	if ev.EventType() == event.EventTypeSelectedWorker {
+		b.reallyRun = true
+	}
+
+	if !b.reallyRun {
+		b.cacheEvents = append(b.cacheEvents, ev)
+		return nil
+	}
+
+	tx, err := b.conn.Begin()
+	if err != nil {
+		return err
+	}
+	defer Rollback(tx)
+
+	if !b.dbInited {
+		err := b.initDbStuff(tx)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = b.saveEvent(tx, ev)
+	if err != nil {
+		return err
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return err
+	}
+	return b.conn.Bus().Notify(buildEventsChannel(b.id))
+}
+
+func (b *inMemoryCheckBuild) Events(from uint) (EventSource, error) {
+	notifier, err := newConditionNotifier(b.conn.Bus(), buildEventsChannel(b.id), func() (bool, error) {
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return newBuildEventSource(
+		b.id,
+		"check_build_events",
+		b.conn,
+		notifier,
+		from,
+		true,
+	), nil
+}
+
+// AbortNotifier returns NoopNotifier because there is no way to abort a in-memory
+// check build. Say a in-memory build may run on ATC-a, but abort-build API call
+// might be received by ATC-b, there is not a channel for ATC-b to tell ATC-a to
+// mark the in-memory build as aborted. If we really want to abort a in-memory
+// check build in future, it might need to add a new table "aborted-in-memory-builds"
+// and API insert in-memory build id to the table, and AbortNotifier watches the
+// table to see if current build should be aborted.
+func (b *inMemoryCheckBuild) AbortNotifier() (Notifier, error) {
+	return newNoopNotifier(), nil
+}
+
+func (b *inMemoryCheckBuild) HasPlan() bool {
+	if b.plan.ID != "" {
+		return true
+	}
+	resource, ok := b.checkable.(Resource)
+	return ok && resource.BuildSummary() != nil && resource.BuildSummary().PublicPlan != nil
+}
+
+func (b *inMemoryCheckBuild) PublicPlan() *json.RawMessage {
+	if b.plan.ID != "" {
+		return b.plan.Public()
+	}
+
+	resource, ok := b.checkable.(Resource)
+	if !ok || resource.BuildSummary() == nil || resource.BuildSummary().PublicPlan == nil {
+		return nil
+	}
+	bytes, err := json.Marshal(resource.BuildSummary().PublicPlan)
+	if err != nil {
+		return nil
+	}
+	m := json.RawMessage(bytes)
+	return &m
+}
+
+func (b *inMemoryCheckBuild) ResourceCacheUser() ResourceCacheUser {
+	return NoUser()
+}
+
+func (b *inMemoryCheckBuild) ContainerOwner(planId atc.PlanID) ContainerOwner {
+	return NewInMemoryCheckBuildContainerOwner(b.ID())
+}
+
+// SaveImageResourceVersion does nothing as a resource check doesn't belong to any job.
```

Comment:
Haven't quite figured out how resource caches work with in-memory builds, but does this mean that we don't keep image resource caches around at all? Does this mean that every time the check runs, we have to fetch the image again?

Reply:
Yes, looks like each in-memory check build needs to fetch image again, but image can be found from local resource cache with the 'cache streamed volume` feature.

I haven't looked into `SaveImageResourceVersion` in depth. Let me revisit it today.

Reply:
@aoldershaw I looked into this function today. In short, this seems not a problem, we don't have to do nothing for in-memory build in this function.

I did a test with a pipeline like below:

```yaml
- resource_types:
  - name: gitlab
    type: registry-image
    source: ..

- resources:
  - name: merge-request
    type: gitlab
    source: ...

- jobs:
  - name: job1
    plan:
    - get: merge_request
    # no trigger, I don't want new version to trigger build automatically
```

With this pipeline, it will only run in-memory check build. Let's see the following tables:

__resource_caches__

|  id  | resource_config_id  | ...  | 
|---|---|---|
| 1   | 1   |  <this is `gitlab`'s image>   | 

__resource_configs__

| id | base_resource_type_id | resource_cache_id | last_referenced |